numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2563,20 +2563,19 @@ object_arrtype_dealloc(PyObject *v)
  */
 
 #define _WORK(num)  \
-    if (type->tp_bases && (PyTuple_GET_SIZE(type->tp_bases)==2)) { \
-        PyTypeObject *sup; \
-        /* We are inheriting from a Python type as well so \
-           give it first dibs on conversion */ \
-        sup = (PyTypeObject *)PyTuple_GET_ITEM(type->tp_bases, num); \
-        /* Prevent recursion */ \
-        if (thisfunc != sup->tp_new) { \
-            robj = sup->tp_new(type, args, kwds); \
-            if (robj != NULL) goto finish;        \
-            if (PyTuple_GET_SIZE(args)!=1) return NULL; \
-            PyErr_Clear(); \
-        } \
-        /* now do default conversion */ \
-    }
+    assert (type->tp_bases && (PyTuple_GET_SIZE(type->tp_bases) == 2)); \
```

Comment:
This assertion is failing for subclasses of `np.float64`. The fact that it is suggests that this code isn't doing what it's supposed to be doing anyway though...

Reply:
This is patently scary... Is there a reason it is not hardcoding `sup` directly?

Reply:
Agreed, very scary - fixed in the follow-up commit.

Reply:
Can remove the `num` completely, as far as I can tell, it basically maps to a known instance?

Reply:
It does indeed map to a known type, but getting hold of that type through the template mechanism is harder than grabbing it at runtime. Essentially you're choosing between:
```python
class A(B, C):
    def __new__(cls, ...):
        C.__new__(...)
```
and
```python
class A(B, C):
    def __new__(cls, ...):
        A.__bases__[1].__new__(...)
```
.

Both are completely correct, unlike what was there before which was outright stupid:
```python
class A(B, C):
    def __new__(cls, ...):
        cls.__bases__[1].__new__(...)
```

Reply:
Yes, the first seemed nicer to me? although not sure if we have the `PyFloatType` name available easily. Could also think about `__bases__[-1]`. in any case, not sure it would make things better or worse, so OK with just leaving it and not doing the last thing is a nice fix.

Reply:
Well, `num` is needed because for `unicode_` and `bytes_` it's `__bases__[0]`, but for `float_` it's `__bases__[1]`.

Reply:
I suspect I will need to make further cleanups here to address #15385 anyway, in future PRs.